backend/src/connectors/fileScanning/clamAv.ts

```diff
@@ -0,0 +1,88 @@
+import NodeClam from 'clamscan'
+import { Readable } from 'stream'
+
+import { getObjectStream } from '../../clients/s3.js'
+import FileModel, { FileInterfaceDoc, ScanState } from '../../models/File.js'
+import log from '../../services/log.js'
+import config from '../../utils/config.js'
+import { ConfigurationError } from '../../utils/error.js'
+import { BaseFileScanningConnector, FileScanResult } from './Base.js'
+
+let av: NodeClam
+const toolName = 'Clam AV'
+
+export class ClamAvFileScanningConnector extends BaseFileScanningConnector {
+  constructor() {
+    super()
+  }
+
+  async init() {
+    if (!av) {
+      try {
+        av = await new NodeClam().init({ clamdscan: config.avScanning.clamdscan })
+      } catch (error) {
+        throw ConfigurationError('Cannot retrieve Clam AV config', {
+          clamAvConfig: config.avScanning,
+        })
+      }
+    }
+  }
+
+  async scan(file: FileInterfaceDoc) {
+    const avStream = av.passthrough()
+    const s3Stream = (await getObjectStream(file.bucket, file.path)).Body as Readable
+    s3Stream.pipe(avStream)
+    log.info({ modelId: file.modelId, fileId: file._id, name: file.name }, 'Scan started.')
+    if (!file.avScan.find((result) => result.toolName === toolName)) {
+      const updatedAvScanArray = [...file.avScan, { toolName: toolName, state: ScanState.InProgress }]
+      file.avScan = updatedAvScanArray
+      await FileModel.updateOne(
```

Comment:
In my opinion, these scanners should not be responsible for storing the scan information in the DB, they should just be an interface with a particular scanning and this method should take in a file stream and return a scan result object. Ther service layer should be responsible for getting the file stream and processing the scan results.

Reply:
This is a bit tricky as it's async and the service that calls these connectors already returns the response to the user. We'd need to call the scanners, wait for the results, store them, _then_ return a 200 response to the user from the backend.

Reply:
Just because we want a process to run asynchronously doesn't mean all the implementation needs to be in a single method. Use a `.then()` on the asynchronous call to provide the implementation for how we will process the results once the scanners are complete.